### PR TITLE
MIR borrowck: print lvalues in error messages in the same way that the AST borrowck

### LINKED ISSUE #44974: MIR-borrowck: print named fields with `.name` rather than `.<num>`
The MIR-borrowck printing of Lvalues visible to end-users prints accesses to a named struct/enum field with a description like `obj.3` instead of `obj.field`.

This needs to be fixed before MIR-borrowck is usable in real-world.

(And fixing it sooner would probably make developing MIR-borrowck a bit more enjoyable.)

### PR DESCRIPTION
**Pull request by zilbuz**
Fix #44974 

- Print fields with `.name` rather than `.<num>`
- Autoderef values if followed by a field or an index
- Output `[..]` when borrowing inside a slice

### CODE REVIEW

**Comment by zilbuz** [comment]
I'd like some feedback on this PR. I can't think of any example that trigger the messages marked `debug!()` in my diff of `borrow_check.rs`. Can someone give me some example ?

**Comment by pnkfelix** [comment]
@zilbuz I'm about to hit the sack, but if you use the following input code, I think you *should* see some examples of some of the cases you are wondering about:

```rust
#![feature(slice_patterns)]

fn main() {
    #[derive(Debug)]
    enum E<X> { A(X), B { x: X } }

    #[derive(Debug)]
    struct S<'a, X: 'a, Y>{ x: &'a [X], y: (Y, Y), };

    let e = E::A(3);

    match e {
        E::A(ref ax) => println!("e.ax: {:?}", ax),
        E::B { x: ref bx } => println!("e.bx: {:?}", bx),
    }

    let s = S { x: &[1, 2, 3,], y: (999, 998) };

    match s {
        S  {
            x: &[ref x0, ref xn..],
            y: (ref y0, ref y1)
        } => {
            println!("s.x0: {:?} s.xn: {:?}", x0, xn);
            println!("s.y0: {:?}, s.y1: {:?}", y0, y1);
        }

        _ => panic!("other case"),
    }
}
```

**Comment by zilbuz** [comment]
r? @pnkfelix 

The only commit untested is https://github.com/rust-lang/rust/pull/44985/commits/e32e81c9da5adcc77bd12192d45d2d4af54d0ee9. I couldn't find a way to access a field through a subslice. I'd have to be able to do something like `[1..3].x` or `xn @ [F { x }..]` in a match maybe...

File: src/test/compile-fail/borrowck/borrowck-describe-lvalue.rs [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+        let mut u = U { b: 0 };
+        let _ra = &mut u.a;
+        u.a; //[ast]~ ERROR cannot use `u.a` because it was mutably borrowed
+             //[mir]~^ ERROR cannot use `u.a` because it was mutably borrowed (Ast)
+             //[mir]~| ERROR cannot use `u.a` because it was mutably borrowed (Mir)
+    }
+    // Static and field from struct
+    unsafe {
+        let _x = sfoo.x();
+        sfoo.x; //[mir]~ ERROR cannot use `sfoo.x` because it was mutably borrowed (Mir)
```
Comment (pnkfelix) [inline src/test/compile-fail/borrowck/borrowck-describe-lvalue.rs]: oh interesting... 

(I think this represents a bug in MIR-borrowck. But fixing that is orthogonal to this PR in any case.)

Comment (pnkfelix) [reply src/test/compile-fail/borrowck/borrowck-describe-lvalue.rs]: (likewise for all the other cases below that are paths based off of static variables)

**Comment by pnkfelix** [comment]
@bors r+

thanks for all the work here @zilbuz !

File: src/test/compile-fail/borrowck/borrowck-describe-lvalue.rs (hunk 1 above)
Comment (pnkfelix) [reply src/test/compile-fail/borrowck/borrowck-describe-lvalue.rs]: Filed as https://github.com/rust-lang/rust/issues/45129

**Comment by kennytm** [comment]
@bors retry 

* Android SDK HTTPS issue (fixed in #45193)
